src/core/layout/qgslayoutaligner.cpp

```diff
@@ -97,6 +97,66 @@ void QgsLayoutAligner::distributeItems( QgsLayout *layout, const QList<QgsLayout
   if ( items.size() < 2 )
     return;
 
+  double minCoord = std::numeric_limits<double>::max();
+  double maxCoord = std::numeric_limits<double>::lowest();
+  QMap< double, QgsLayoutItem * > itemCoords;
+
+  double step;
+
+  // equispaced distribution doesn't follow the same approach of the other distribution types
+  if ( distribution == DistributeHSpace || distribution == DistributeVSpace )
+  {
+    double length = 0.0;
+
+    for ( QgsLayoutItem *item : items )
+    {
+      QRectF itemBBox = item->sceneBoundingRect();
+      double item_min, item_max;
+
+      item_min = ( distribution == DistributeHSpace ? itemBBox.left() :
+                   itemBBox.top() );
+      item_max = ( distribution == DistributeHSpace ? itemBBox.right() :
+                   itemBBox.bottom() );
+
+      minCoord = std::min( minCoord, item_min );
+      maxCoord = std::max( maxCoord, item_max );
+      length += ( item_max - item_min );
+      itemCoords.insert( item_min, item );
+    }
+    step = ( maxCoord - minCoord - length ) / ( items.size() - 1 );
+
+    double currentVal = minCoord;
+    layout->undoStack()->beginMacro( undoText( distribution ) );
+    for ( auto itemIt = itemCoords.constBegin(); itemIt != itemCoords.constEnd(); ++itemIt )
+    {
+      QgsLayoutItem *item = itemIt.value();
+      QPointF shifted = item->pos();
+
+      layout->undoStack()->beginCommand( itemIt.value(), QString() );
+
+      if ( distribution == DistributeHSpace )
+      {
+        shifted.setX( currentVal );
+      }
+      else
+      {
+        shifted.setY( currentVal );
+      }
+
+      QgsLayoutPoint newPos = layout->convertFromLayoutUnits( shifted, item->positionWithUnits().units() );
+      item->attemptMove( newPos );
+
+      layout->undoStack()->endCommand();
+
+      currentVal += ( distribution == DistributeHSpace ? item->rect().width() :
+                      item->rect().height() ) + step;
+
+      layout->undoStack()->endMacro();
```

Comment:
That should probably be out of the loop one line down

Reply:
Good catch, wrong placement.